### ISSUE TITLE
Tridactyl doesn't work on Google Docs

### ISSUE THREAD

**Issue by cprn** [tridactyl/tridactyl#1971]
Tridactyl version: 1.16.2pre2999-95db0a8
Firefox version: Mozilla Firefox 70.0
Operating system: linux

*   Brief description of the problem:

When I edit any public Google Docs document Tridactyl stays in `normal` mode even though I'm editing and actively inserting characters in the document. Since it's in `normal` mode all the time I can't use escape to go to `normal` mode and use any of the Tridactyl features.

*   Steps to reproduce:

1. Open Firefox,
2. go to e.g. [this document](https://docs.google.com/document/d/1OxHtRDS65d6Te2mbzCjNXzjFjeTdm8gAgPFCuwGENcg/edit),
3. type something, notice you're typing, notice you're in `normal` mode,
4. try to press `f` to go activate links following, notice it doesn't work.

*   URL of the website the bug happens on:
https://docs.google.com

*   Config (in a new tab, run `:viewconfig`, copy the url and paste it somewhere like gist.github.com):
https://gist.github.com/cprn/2df156c5fba10ff41a17cc82c9f07a3c

*   Contents of ~/.tridactylrc or ~/.config/tridactyl/tridactylrc (if they exist):

```sh
$ ls .tridactyl .config/tridactyl
ls: cannot access '.tridactyl': No such file or directory
ls: cannot access '.config/tridactyl': No such file or directory
```

**Comment by bovine3dom** [comment]
You're not really in normal mode. (For the record, Escape in normal mode is bound to "enter normal mode": `:bind <Esc> # <Escape> = composite mode normal ; hidecmdline` anyway.) I think you're actually in insert mode but our mode indicator is a bit buggy (`<C-6>` works for switching back to the previous tab, for example).

The reason `<Escape>` doesn't get intercepted by Tridactyl in Google docs is beyond me, but that's probably the real bug :)

**Comment by cprn** [comment]
I edited the title but I can't confirm `<c-6>` works, it doesn't for me. As soon as I go to tab with Google Docs document I'm trapped... Some Firefox shortcuts like `<c-f>` and `<c-K>` don't work either. Google seems javascript hungry for keyboard events.

**Comment by shioyama** [comment]
Arg this is a killer, I keep accidentally inserting characters into google documents... I wish there was a workaround, even a hacky one.

**Comment by csakoda** [comment]
Anyone find a workaround for this?

**Comment by bovine3dom** [comment]
I think lots of people put Tridactyl in ignore mode on Google Docs to train themselves out of trying to use it: `:autocmd DocStart ^https://docs\.google\.com/ mode ignore`

**Comment by mpgarate** [comment]
I am looking for a workaround for this as well. There are two behaviors I am hoping to resolve:
1) let "esc" work on google docs to enable normal mode
2) keep normal mode active when switching into a google doc tab, until I explicitly enter insert mode. If possible I would actually like to make that setting global for all sites.

For number 2, I found that clicking with the mouse in the top section of the google doc restores normal mode, but I was not able to find a way to do that programmatically. I tried running some js on DocStart/DocLoad, but it was blocked by the content security policy.

**Comment by bovine3dom** [comment]
@mpgarate in Tridactyl Beta you could try `set blacklistkeys / ,` and then ctrl-, (comma) might do what you want escape to. (This is a guess and it may not work).

For 2 have you tried `set allowautofocus false`? There is also preventautofocusjackhammer but that is extreme.

**Comment by bovine3dom** [comment]
> I tried running some js on DocStart/DocLoad, but it was blocked by the content security policy.

I don't think this is right - that error pops up a lot because of Tridactyl anyway. Your js probably didn't work for some other reason :)

**Comment by mpgarate** [comment]
Thanks for the suggestions. I tried a few variations of the following but was not able to change the behavior. I tested by using J and K to switch to and from a tab with a Google Doc. I tested loading the url in a fresh tab and trying to switch away, as well as loading the page, clicking a different tab, and then switching in and out with J and K. In each case I found that the only way to switch out of the doc using J or K was by using the mouse to first click outside of the input area. If I had done that at least once, it would be remembered the next time I switched back to the tab.

```
set allowautofocus false                                                                                                                                                                                         
seturl ^https://docs.google.com/ preventautofocusjackhammer true
```

**Comment by bovine3dom** [comment]
Ah, it's probably an iframe issue. Farnoy is trying to fix that in #2693 but so far we haven't had any luck with it :)

**Comment by iamchanghao** [comment]
It seems like still no resolution to this?
I use a temporary annoying workaround to go around this issue, if you press <ctrl-F> it will call the native search window on google doc, then press <tab> to move away from the search box, then the Tridactyl key press is enabled

**Comment by BlueFalconHD** [comment]
The reason for this is the system google uses. This happens on all google drive sites because google uses custom keybinds and right click menus, so it tells the browser to ignore them.

**Comment by basepi** [comment]
Escape hatch (`<C-,>`) does seem to work for me, but then I struggle to re-focus the input without going to the mouse. `gi` doesn't do it. I've taken to using `f` to hint some of the chrome, "click" on my profile picture, and then hit `esc`, which puts the focus back in the doc.

**Comment by bovine3dom** [comment]
You might find the `;;` hint mode lets you get back to the page without actually having to "click" anything

**Comment by basepi** [comment]
I don't think so. It seems like Tridactyl just has no idea the document frame exists:

<img width="1225" alt="image" src="https://github.com/tridactyl/tridactyl/assets/702318/4311fa55-778e-4d0c-b523-da130ac4a310">

Obviously, I haven't tried every one of those hints, but I tried a couple, and none of them get my cursor flashing again.

**Comment by basepi** [comment]
Ooh, I did find a better way. If I just enter ex mode (`:`) and then hit <Esc>, focus goes back to the cursor.

**Comment by mpgarate** [comment]
> Escape hatch (<C-,>) does seem to work for me

Is this still working for you / anyone else? I am not able to leave focus from the text input area

**Comment by basepi** [comment]
Yes, it works for me. Pauses the flashing cursor and lets me navigate between tabs with J and K, which is my primary concern.

**Comment by mpgarate** [comment]
Thank you for confirming- I may be missing a step here.

In a Google docs tab, for me the cursor selects the text area by default. If I press Control + `,` it changes the input text to a subscript style, leaving the input area selected.

For example if I load a new google doc and type: `foo bar<C-,>JK` to try to switch tabs, I see:

![image](https://github.com/tridactyl/tridactyl/assets/476374/c3821a73-a582-4bfc-b773-c95589c96cdf)

**Comment by basepi** [comment]
Ah, this is probably because I'm on a mac. So `<Cmd>+,` is the subscript keybind, leaving `<Ctrl>+,` to work for escape hatch. Not sure how to solve this for windows.

**Comment by bovine3dom** [comment]
In a similar vein, you can run `:bind --mode=browser <A-,> escapehatch` and then use alt+, on Google Docs (or maybe everywhere). And pray that Google doesn't decide to steal that bind from us too :)

**Comment by cprn** [comment]
I can confirm what @mpgarate said - `ctrl+,` just changes cursor to subscript. Binding to `alt+,` works but using a different binding for one website is hardly a solution. Still, better than nothing.

**Comment by bovine3dom** [comment]
Fixed in #5381
